designs/odcr.md

```diff
@@ -0,0 +1,140 @@
+
+## Goals
+
+- Support associating ODCR to EC2NodeClass
+- Define Karpenter's behavior when launching nodes into Capacity Reservation
+- Define Karpenter's behavior when encountering errors when attempting to launch nodes into Capacity Reservation
+- Define Karpenter's behavior when consolidating nodes in Capacit Reservation
+
+## Non-Goals
+
+_We are keeping the scope of this design very targeted so even if these could be things we eventually support, we aren't scoping them into this design_
+- Supporting prioritization when launching nodes. We will delegate this to weights within NodePools and [default behavior](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet-on-demand-backup.html#ec2-fleet-on-demand-capacity-reservations) of the CreateFleet API. 
+- Supporting changes in scaling behavior when ODCR is associated to a NodeClass. _We won't bring up N nodes to match an N node capacity reservation_
+
+## Proposed Solution
+
+### Supporting associating ODCR to EC2NodeClass
+
+Add a new field `capacityReservationSpec` to `EC2NodeClass` 
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  capacityReservationSpec:
+    capacityReservationPreference: open | none | None  # Cannot be defined if capacityReservationTarget is specified
+    capacityReservationTarget: # Cannot be defined if capacityReservationPreference is specified
+      capacityReservationId: String | None
+      capacityReservationResourceGroupArn: String | None
+```
+This follows exactly how LaunchTemplate defines [CapacityReservationSpecification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationspecification.html) with the same
+constraints. 
+
+Karpenter will perform validation against the spec to ensure there isn't any violation prior to creating the LaunchTemplates.
+
+### Launching Nodes into Capacity Reservation
+
+#### All Launch Templates are associated with the specified Capacity Reservation
+
+EC2NodeClass currently supports automatically generating Launch Templates based on instance types and their AMI requirements. We won't implement any prioritization within Karpenter for Capacity Reservation and instead rely on existing behavior of CreateFleet and Capacity Reservation. Specifically, CreateFleet when creating instances against a single targeted Capacity Reservation will fail if the Reservation is fully utilized.
+However in the case where a Capacity Reservation Group is used it will fall back into On Demand nodes when the Reservation Group limit is hit. 
+
+We won't introduce any other prioritization nor fallback when launching nodes against EC2NodeClasses with a Capacity Reservation. This means that it is the user's responsibility to ensure the Capacity Reservation associated to a EC2NodeClasses will include instance types, availability zone and platform the EC2NodeClasses should launch.
```

Comment:
💯  - the UX would be better if we performed the few additional API calls - when Karpenter sees a new reservation for the first time, it can query the AZ its allocated within, the instance type, etc. which will improve the UX for users